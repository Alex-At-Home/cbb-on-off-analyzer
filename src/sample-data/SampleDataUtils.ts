import _ from "lodash";

export class SampleDataUtils {

  /** For snapshot testing, provides a more useful output than a giant list of stats */
  static summarizeEnrichedApiResponse(exampleRawResponse: Record<string, any>, mutableSample: boolean = false) {
    const expectedFields1 = mutableSample ? null : new Set(_.keys(exampleRawResponse));

    return {
      test: (val: any) => val && (val.off_efg ||
          val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg") ||
          val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")
        ),
        //(pick a field that all stat sets have - plus special case for: diagnostics)
      print: (val: any, serialize: ((val: any) => string), indent: ((val: any) => string)) => {
        const expectedFields = expectedFields1 || new Set(_.keys(exampleRawResponse));
        if (val.off_efg) {
          return "{\n" +
            _.chain(val).toPairs().filter(kv => {
              return !expectedFields.has(kv[0]) //derived field)
                || (!kv[1]?.hasOwnProperty("value") && exampleRawResponse[kv[0]]?.hasOwnProperty("value")) //missing value
                || (kv[1]?.hasOwnProperty("value") && _.keys(kv[1]).length > 1) //(has extra properties)
            }).sortBy(0).map(kv => {
              return indent(`${kv[0]}: ` + serialize(kv[1]))
            }).value().join("\n\n") +
          "}";
        } else if (val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg")) {
          return JSON.stringify(_.pick(val, [
            "oRtg", "adjORtg", "adjORtgPlus", "dRtg", "adjDRtg", "adjDRtgPlus"
          ]), null , 3);
        } else if (val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")) {
          return JSON.stringify(_.pick(val, [
            "deltaOffPpp", "deltaOffEfg", "deltaOffAdjEff", "deltaDefPpp", "deltaDefEfg", "deltaDefAdjEff"
          ]), null , 3);
        } else {
          return "???";
        }
      }
    }
  }


  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateLineup(input: Record<string, any>) {
    const lineupAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.toPairs(lineupAsObj).map(kv => {
      return { key: kv[0], ...(kv[1] as Record<string, any>) };
    })
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateTeam(input: Record<string, any>) {
    const teamAsObj = _.transform(_.toPairs(input), (acc, inElKv) => {
      _.forEach(_.toPairs(inElKv[1] as Record<string, any>), (kv) => {
        _.forEach(_.toPairs(kv[1] as Record<string, any>), (deepKv) => {
          if (inElKv[0] != "main") {
            _.set(acc, `${kv[0]}.${deepKv[0]}${inElKv[0]}.value`, deepKv[1]);
          } else {
            _.set(acc, `${deepKv[0]}.${kv[0]}`, deepKv[1]);
          }
        });
      });
    }, {} as Record<string, any>);
    return teamAsObj;
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplatePlayer(input: Record<string, any>) {
    const flatPlayersAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.transform(["baseline", "on", "off"], (acc, v) => {
      acc[v] = { player: {
        buckets: _.chain(flatPlayersAsObj).toPairs().filter(kv => {
          return kv[0].indexOf(v) == 0;
        }).map(kv => {
          const key = kv[0].substring(v.length + 2); //(+2 for ": ")
          return { ...(kv[1]), key: key }
        }).value()
      } };
    }, {} as Record<string, any>);
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseTeam(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().map((kv: [string, any]) => {
        return { key: kv[0], ...(kv[1]) }
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponsePlayer(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().flatMap((kv: [string, any]) => {
        const players = (kv[1]?.player?.buckets || []) as Array<Record<string, any>>;
        return players.map(p => { return { ...p, key: `${kv[0]}: ${p.key}` } });
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseLineup(input: Array<Record<string, any>>, filter?: Set<String>) {
    const preSorted = _.transform(input, (acc, inEl) => {
      const recordId = inEl?.key || "unknown";
      const injectInfo = (k: string, prefix: string) => {
        const topLevelKey = k.substring(prefix.length);
        const realPrefix = prefix || "main";
        if (prefix) {
          const existingObj = _.get(acc, `${topLevelKey}.${recordId}`) || {};
          existingObj[realPrefix] = inEl[k]?.value;
          const sortedObj = _(existingObj).toPairs().sortBy(0).reverse().fromPairs().value()
          _.set(acc, `${topLevelKey}.${recordId}`, sortedObj);
        } else {
          _.set(acc, `main.${topLevelKey}.${recordId}`, inEl[k]);
        }
      };

      if (!filter || filter.has(recordId)) {
        _.keys(inEl).filter(k => k != "key").forEach(k => {
          const stat = _.find([
            "oppo_total_off_", "oppo_total_def_",
            "oppo_off_", "oppo_def_",

            "team_total_off_", "team_total_def_",
            "team_off_", "team_def_",

            "total_off_", "total_def_",
            "off_", "def_"
          ], prefix => k.indexOf(prefix) == 0);
          if (stat) {
            injectInfo(k, stat);
          } else {
            injectInfo(k, "");
          }
        });
      }
    }, {} as Record<string, any>);

    // Sort the keys into some sort of logical order
    return _.chain(preSorted).toPairs().sortBy([
      (kv: [ string , any ]) => {
        const key = kv[0];
        if (key == "main") {
          return 100;
        } else if ((key.indexOf("assist") >= 0) || (key.indexOf("to") >= 0) || (key.indexOf("ast") >= 0)) {
          return 4;
        } else if ((key.indexOf("ft") == 0) || (key.indexOf("fg") == 0) || (key == "efg")) {
          return 1;
        } else if ((key.indexOf("2p") == 0) || (key.indexOf("3p") == 0)) {
          return 2;
        } else if ((key.indexOf("rb") > 0)) {
          return 3;
        } else if ((key.indexOf("adj") >= 0)) {
          return 5;
        } else { //misc other stuff
          return 6;
        }
      },
      (kv: [ string , any ]) => {
        return kv[0];
      },
    ]).fromPairs().value();
  }
}
