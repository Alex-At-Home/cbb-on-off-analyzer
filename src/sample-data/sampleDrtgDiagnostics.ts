export const sampleDrtgDiagnostics = {
  "stl": 30,
  "blk": 5,
  "drb": 94,
  "pfPct": 0.2,
  "playerRtg": 142.82324406037372,
  "playerDelta": 8.134297736197055,
  "scPossConceded": 0.6341530662717683,
  "noShotCredit": 32.188268933121634,
  "reboundCredit": 36.07039815349106,
  "missFtCredit": 6.070287539936104,
  "stopsIndPct": 0.21044437889736348,
  "stopsTeamPct": 0.15540255483086815,
  "teamBlk": 60,
  "oppoPts": 1804,
  "oppoPoss": 1766,
  "oppoFga": 1429,
  "oppoFgm": 534,
  "oppoFtm": 426,
  "oppoFta": 626,
  "oppoFtPoss": 297.34999999999997,
  "oppoTov": 59,
  "teamStl": 150,
  "opponentOrbPct": 0.2708757637474542,
  "opponentFgPct": 0.3736878936319104,
  "teamOrbCreditToDefender": 0.2724649000411893,
  "teamOrbCreditToRebounder": 0.16965277015673305,
  "teamDvsRebCredit": 0.616272360069244,
  "oppoFgMiss": 835,
  "oppoNonStlTov": -91,
  "teamMissWeight": 0.4376537866243272,
  "oppoFtPct": 0.6805111821086262,
  "oppoFtHitOnePlus": 266.99856230031946,
  "oppoProbFtHitOnePlus": 0.8979268952423726,
  "oppoScPoss": 800.9985623003195,
  "oppoPtsPerScore": 2.2521888114496065,
  "teamRtg": 102.15175537938845,
  "adjDRtg": 102.04282992219363,
  "adjDRtgPlus": 0.4085659844387266,
  "avgEff": 100,
  "dRtg": 110.2860531155855,
  "offSos": 108.0781993204983
 };
