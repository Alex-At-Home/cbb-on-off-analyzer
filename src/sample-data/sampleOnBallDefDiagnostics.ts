/** Manually built by copying the snapshot from src/utils/stats/__tests__/RatingUtils.test.ts */
export const sampleOnBallDefDiagnostics =
{
   "playerRebCredit": 0.09664054373332856,
   "teamRebCredit": 0.14616882239665946,
   "comboRebCredit": 0.2948067270155614,
   "playerTargetPoss": 185.586,
   "targetedPct": 0.12134831460674157,
   "playerPtsPerScore": 2.4291266575217194,
   "onBallStopCredit": 0.4668039520787198,
   "offBallPoss": 1572.7865168539327,
   "offBallPts": 1539.0286760055637,
   "otherRebounds": 631.6805699093197,
   "otherRebCredit": 0.14474344317512738,
   "offBallStopCredit": 0.39516433361719344,
   "onVsOffBallWeight": 0.4,
   "weightedPtsPerScore": 2.126826397164934,
   "onBallCreditWeight": 0.04853932584269663,
   "offBallCreditWeight": 0.13179775280898876,
   "comboBallStopCredit": 0.8290444280732188,
   "unadjDRtg": 103.9658869175629,
   "weightedClassicDRtgMean": 98.84955759282161,
   "weightedUnadjDRtgMean": 107.19011038664136,
   "uncategorizedAdjustment": -7.10724422350558,
   "adjustedPossPct": 0.08000000000000002,
   "dRtg": 96.85864269405732,
   "adjDRtg": 89.6509562344713,
   "adjDRtgPlus": -2.069808753105741
};
