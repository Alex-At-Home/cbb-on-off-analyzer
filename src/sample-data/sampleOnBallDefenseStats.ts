export const sampleOnBallDefenseStats =
[
  {
     "code": "totals",
     "title": "Team Name",
     "pts": 1554,
     "plays": 1780,
     "scorePct": 39.1,
     "tovPct": 16.5,
     "fgMiss": 769,
     "totalPts": -1,
     "totalScorePct": -1,
     "totalPlays": -1,
     "uncatPts": -1,
     "uncatPlays": -1,
     "uncatScorePct": -1,
     "uncatPtsPerScPlay": -1
  },  [
   {
      "code": "Code1",
      "title": "#1 Player1",
      "pts": 170,
      "plays": 216,
      "scorePct": 32.4,
      "tovPct": 15.7,
      "fgMiss": 111,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code2",
      "title": "#02 Player2",
      "pts": 159,
      "plays": 170,
      "scorePct": 41.2,
      "tovPct": 11.8,
      "fgMiss": 79,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code3",
      "title": "#13 Player3",
      "pts": 151,
      "plays": 169,
      "scorePct": 37.3,
      "tovPct": 16,
      "fgMiss": 78,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code4",
      "title": "#99 Player4",
      "pts": 109,
      "plays": 107,
      "scorePct": 39.3,
      "tovPct": 12.1,
      "fgMiss": 52,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code5",
      "title": "#21 Player5",
      "pts": 103,
      "plays": 148,
      "scorePct": 31.8,
      "tovPct": 14.9,
      "fgMiss": 79,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code6",
      "title": "#66 Player6",
      "pts": 52,
      "plays": 88,
      "scorePct": 26.1,
      "tovPct": 11.4,
      "fgMiss": 54,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   },
   {
      "code": "Code7",
      "title": "Player 7",
      "pts": 76,
      "plays": 95,
      "scorePct": 32.6,
      "tovPct": 14.7,
      "fgMiss": 50,
      "totalPts": -1,
      "totalScorePct": -1,
      "totalPlays": -1,
      "uncatPts": -1,
      "uncatPlays": -1,
      "uncatScorePct": -1,
      "uncatPtsPerScPlay": -1
   } ]
];
