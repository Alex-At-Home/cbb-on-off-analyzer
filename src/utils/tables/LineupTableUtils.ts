// Lodash
import _ from "lodash";

// Util imports
import { StatModels, PlayerCodeId, PlayerCode, PlayerId, Statistic, IndivStatSet, TeamStatSet, LineupStatSet, IndivPosInfo, RosterEntry } from '../StatModels';
import { RatingUtils, OnBallDefenseModel } from "../stats/RatingUtils";
import { PositionUtils } from "../stats/PositionUtils";
import { LineupUtils } from "../stats/LineupUtils";
import { LuckUtils, OffLuckAdjustmentDiags, DefLuckAdjustmentDiags, LuckAdjustmentBaseline } from "../stats/LuckUtils";
import { ParamDefaults } from '../FilterModels';

export type PositionInfo = PlayerCodeId & { numPoss: number };

/** Object marshalling logic for lineup tables */
export class LineupTableUtils {

  /** Key of "total" fake lineuo */
  static readonly totalLineupId = "TOTAL";

  /** Handy accessor for picking the player codes out of the lineup */
  static buildCodesAndIds(lineup: LineupStatSet): Array<PlayerCodeId> {
    return lineup.players_array ?
      (lineup.players_array?.hits?.hits?.[0]?._source?.players || []) :
      _.toPairs((lineup.player_info || {}) as Record<PlayerId, IndivStatSet>).map(kv => { return { code: kv[1].code, id: kv[0] } }) //(leaderboard mode)
      ;
  }

  /** Injects some advanced stats into players, returns an associative array vs player.key */
  static buildBaselinePlayerInfo(
    players: Array<IndivStatSet> | undefined,
    globalRosterStatsByCode: Record<PlayerCode, IndivStatSet>, teamStat: TeamStatSet,
    avgEfficiency: number, adjustForLuck: boolean, luckConfigBase: "baseline" | "season",
    onBallDefenseByCode: Record<PlayerCode, OnBallDefenseModel>  = {}
  ): Record<PlayerId, IndivStatSet> {
    const sampleRosterByCode = _.fromPairs( // Needed for ORtg, also ensure the codes exist
      (players || []).map((mutableP: IndivStatSet) => {
        // Code:
        mutableP.code = (mutableP.player_array?.hits?.hits?.[0]?._source?.player?.code || mutableP.key) as PlayerCode;
        return [ mutableP.code, mutableP];
      })
    );

    const baselinePlayerInfo = _.fromPairs(
      (players || []).map((mutableP: IndivStatSet) => {
        const playerAdjustForLuckOff = adjustForLuck; //(only apply luck to the lineups if we have granular lineup 3PA info, but always apply to priors)
        const playerAdjustForLuckDef = adjustForLuck; //(never apply luck to the lineups, but always apply to priors)

        // Possession %
        mutableP.off_team_poss_pct = { value: _.min([(mutableP.off_team_poss.value || 0)
            / (teamStat.off_poss?.value || 1), 1 ]) };
        mutableP.def_team_poss_pct = { value: _.min([(mutableP.def_team_poss.value || 0)
            / (teamStat.def_poss?.value || 1), 1 ]) };

        if (mutableP?.doc_count) {
          const globalPlayerStats = (luckConfigBase == "season") ? 
            (globalRosterStatsByCode[mutableP.code || "??"] || mutableP) : mutableP;
          // (No offensive luck since our "to adjust" and baseline are the same)
          const offLuckAdj = LuckUtils.calcOffPlayerLuckAdj(
            mutableP, globalPlayerStats, avgEfficiency
          );
          // Calculate luck for defense - over the baseline query, but will regress to opponent SoS
          const defLuckAdj = LuckUtils.calcDefPlayerLuckAdj(
            mutableP, globalPlayerStats, avgEfficiency
          );
          LuckUtils.injectLuck(mutableP, offLuckAdj, defLuckAdj);
        }

        // Add ORtg to lineup stats:
        const [ oRtg, adjORtg, rawORtg, rawAdjORtg, oRtgDiag ] = RatingUtils.buildORtg(
          mutableP, sampleRosterByCode, {
              total_off_to: teamStat.total_off_to || { value: 0 },
              sum_total_off_to: { //(sum of all players TOs, so we can calc team TOVs)
                //(note don't luck adjust these since the team values aren't luck adjusted)
                value: _.sumBy(players, p => p.total_off_to?.value || 0)
              }
          }, avgEfficiency, true, playerAdjustForLuckOff
        );
        const [ dRtg, adjDRtg, rawDRtg, rawAdjDRtg, dRtgDiag ] = RatingUtils.buildDRtg(
          mutableP, avgEfficiency, !_.isEmpty(onBallDefenseByCode), playerAdjustForLuckDef
        );
        mutableP.off_rtg = {
          value: oRtg?.value, old_value: rawORtg?.value,
          override: playerAdjustForLuckOff ? "Luck adjusted" : undefined
        };
        mutableP.off_adj_rtg = {
          value: adjORtg?.value, old_value: rawAdjORtg?.value,
          override: playerAdjustForLuckOff ? "Luck adjusted" : undefined
        };
        mutableP.off_usage = {
          value: !_.isNil(oRtgDiag) ? oRtgDiag.Usage!*0.01 : mutableP.off_usage?.value
        };
        mutableP.def_rtg = {
          value: dRtg?.value, old_value: rawDRtg?.value,
          override: playerAdjustForLuckDef ? "Luck adjusted" : undefined
        };
        mutableP.def_adj_rtg = {
          value: adjDRtg?.value, old_value: rawAdjDRtg?.value,
          override: playerAdjustForLuckDef ? "Luck adjusted" : undefined
        };

        // Apply on-ball defense if it exists for this player
        if (dRtgDiag && onBallDefenseByCode.hasOwnProperty(mutableP.code!)) {
          const onBallDefense = onBallDefenseByCode[mutableP.code!]!;
          const onBallDiags = RatingUtils.buildOnBallDefenseAdjustmentsPhase1(mutableP, dRtgDiag, onBallDefense);
          dRtgDiag.onBallDef = onBallDefense;
          dRtgDiag.onBallDiags = onBallDiags;
        
          mutableP.diag_def_rtg = dRtgDiag;
        }

        // If roster info is available then add:
        const rosterEntry = globalRosterStatsByCode[mutableP.code!].roster;
        if (rosterEntry && !_.isEmpty(rosterEntry)) {
          mutableP.roster = rosterEntry;
        }

        return [ mutableP.key, mutableP ];
      })
    );

    // Finish off on-ball defense if there is any:
    if (!_.isEmpty(onBallDefenseByCode)) {
      RatingUtils.injectOnBallDefenseAdjustmentsPhase2(_.values(baselinePlayerInfo), teamStat);
    }
    return baselinePlayerInfo;
  }

  /** Handy util for sorting JSON blobs of fields */
  static sorter(sortStr: string) { // format: (asc|desc):(off_|def_|diff_)<field>|year
    const sortComps = sortStr.split(":"); //asc/desc
    const dir = (sortComps[0] == "desc") ? -1 : 1;
    const fieldComps = _.split(sortComps[1], "_", 1); //off/def/diff
    const fieldName = (fieldComps[0] != "year")
      ? sortComps[1].substring(fieldComps[0].length + 1) //+1 for _
      : sortComps[1];
    const field = (stat: any) => {
      switch(fieldComps[0]) {
        case "diff": //(off-def)
          return (stat["off_" + fieldName]?.value || 0.0)
                - (stat["def_" + fieldName]?.value || 0.0);
        case "year": //metadata
          return parseInt((stat?.year || ParamDefaults.defaultLeaderboardYear).substring(0, 4));
        default:
          return stat[sortComps[1]]?.value; //(off or def)
      }
    };
    return (stat: any) => {
      return dir*(field(stat) || 0);
    };
  };


  /** Builds positional info vs player key */
  static buildPositionPlayerMap(
    players: IndivStatSet[] | undefined, teamSeasonLookup: string, externalRoster?: Record<PlayerId, RosterEntry>
  ): Record<PlayerId, IndivPosInfo> {
    const positionFromPlayerKey = _.chain(players || []).map((player: IndivStatSet) => {
      const rosterMeta = player.roster || externalRoster?.[player.key];
      const [ posConfs, posConfsDiags ] = PositionUtils.buildPositionConfidences(player, rosterMeta?.height_in);
      const [ pos, posDiags ] = PositionUtils.buildPosition(posConfs, posConfsDiags.confsNoHeight, player, teamSeasonLookup);

      return [ player.key, { 
        posConfidences: _.values(posConfs || {}), 
        posClass: pos,
        roster: rosterMeta ? {
          number: rosterMeta.number,
          height: rosterMeta.height,
          year_class: rosterMeta.year_class,
          pos: rosterMeta.pos,
        } : undefined
      } ];
    }).fromPairs().value();

    return positionFromPlayerKey;
  }

  /** Builds a filtered sorted list of lineups */
  static buildFilteredLineups(
    lineups: any[],
    filterStr: string, sortBy: string | undefined, minPoss: string, maxTableSize: string,
    teamSeasonLookup: string | undefined, positionFromPlayerKey: Record<string, any> | undefined,
  ) {
    const [
      filterFragmentsPve, filterFragmentsNve, filterOnPosition
    ] = PositionUtils.buildPositionalAwareFilter(filterStr);

    const filteredLineups = _.chain(lineups).filter((lineup) => {
      const minPossInt = parseInt(minPoss);
      const offPos = lineup.off_poss?.value || 0;
      const defPos = lineup.def_poss?.value || 0;
      return offPos >= minPossInt || defPos >= minPossInt; //(unclear which of || vs && is best...)
    }).filter((lineup) => {

      const codesAndIds = LineupTableUtils.buildCodesAndIds(lineup);

      const lineupPosFromPlayerKey = positionFromPlayerKey || lineup.player_info; //(leaderboard version, calc from lineup)
      const lineupTeamSeason = teamSeasonLookup ||
        `${lineup.gender}_${lineup.team}_${lineup.year}`; //(leaderboard version, calc from lineup)

      const namesToTest = filterOnPosition ?
        PositionUtils.orderLineup(codesAndIds, lineupPosFromPlayerKey, lineupTeamSeason) : codesAndIds;
      const teamFilter = lineup.team ? [ { id: `${lineup.team}_${lineup.year}`, code: lineup.team } ] : []; //(leaderboard version)

      const playerFilter = PositionUtils.testPositionalAwareFilter(
        namesToTest.concat(teamFilter), filterFragmentsPve, filterFragmentsNve
      );

     return playerFilter && (lineup.key != ""); // (workaround for #53 pending fix)
    }).sortBy(
       sortBy ? [ LineupTableUtils.sorter(sortBy) ] : [] //(don't sort if sortBy not specified)
    ).take(
      parseInt(maxTableSize)
    ).value();

    return filteredLineups;
  }

  /** Builds a filtered sorted list of lineups */
  static buildEnrichedLineups(
    // Stats inputs:
    filteredLineups: Array<LineupStatSet>, globalTeamStats: TeamStatSet,
    players: Array<IndivStatSet>, baselineTeamStats: TeamStatSet,
    // Table control:
    adjustForLuck: boolean, luckConfigBase: "baseline" | "season", avgEfficiency: number,
    // Derived objects:
    showTotalLineups: boolean, teamSeasonLookup: string,
    positionFromPlayerKey: Record<PlayerId, any>, baselinePlayerInfo: Record<PlayerId, IndivStatSet>
  ): Array<LineupStatSet> {
    // The luck baseline can either be the user-selecteed baseline or the entire season
    const baseLuckBuilder: () => [TeamStatSet, Record<PlayerId, IndivStatSet>] = () => {
      if (adjustForLuck) {
        switch (luckConfigBase) {
          case "baseline":
            return [
              baselineTeamStats, baselinePlayerInfo
            ];
          default: //("season")
            return [
              globalTeamStats, _.fromPairs((players || []).map(p => [ p.key, p ]))
            ];
        }
      } else return [ StatModels.emptyTeam(), {} ]; //(not used)
    };
    const [ baseOrSeasonTeamStats, baseOrSeason3PMap ] = baseLuckBuilder();

    /** Perform enrichment on each lineup, including luck adjustment */
    const enrichLineup = (lineup: LineupStatSet) => {
      const codesAndIds = LineupTableUtils.buildCodesAndIds(lineup);

      const sortedCodesAndIds = (lineup.key == LineupTableUtils.totalLineupId) ? undefined :
        PositionUtils.orderLineup(codesAndIds, positionFromPlayerKey, teamSeasonLookup);

      const perLineupPlayerLuckMap: Record<PlayerId, IndivStatSet> = _.fromPairs(
        codesAndIds.map((cid: PlayerCodeId) => {
          return [  cid.id, baseOrSeason3PMap[cid.id] || StatModels.emptyIndiv()];
        })
      );
      const luckAdj = (
        (lineup.key != LineupTableUtils.totalLineupId) && adjustForLuck && lineup?.doc_count
      ) ? [
        LuckUtils.calcOffTeamLuckAdj(
          lineup, codesAndIds.map(cid => perLineupPlayerLuckMap[cid.id]), baseOrSeasonTeamStats, perLineupPlayerLuckMap, avgEfficiency,
          baselineTeamStats?.total_off_3p_attempts?.value  
            //(ensure that the aggregation of the 3P-luck-adjusted lineups are equal to the 3P-adjusted set)
        ),        
        LuckUtils.calcDefTeamLuckAdj(
          lineup, baseOrSeasonTeamStats, avgEfficiency,
          baselineTeamStats?.total_def_3p_attempts?.value
            //(ensure that the aggregation of the 3P-luck-adjusted lineups are equal to the 3P-adjusted set)
        ),
      ] as [OffLuckAdjustmentDiags, DefLuckAdjustmentDiags] : undefined;

      if (lineup?.doc_count) {
        LineupUtils.buildEfficiencyMargins(lineup); //(just used for display in the lineup table)

        if (lineup.key != LineupTableUtils.totalLineupId) {
          // don't inject luck into the total lineup calcs - a) empirically it's wrong so
          // probably I'm doing something stupid, b) it seems wrong anyway, eg aggregating
          // the per-lineup players is not correct, (for team we have the actual 3PA numbers)
          // but it's better than assuming all players play in their season poss ratios, which
          // is what this appears to do...
          // (https://github.com/Alex-At-Home/cbb-on-off-analyzer/issues/100)
          LuckUtils.injectLuck(lineup, luckAdj?.[0], luckAdj?.[1]);
          lineup.off_luck_diags = luckAdj?.[0];
          lineup.def_luck_diags = luckAdj?.[1];
        }
      }
      return lineup;
    };

    const enrichedLineups = filteredLineups.map(lineup => enrichLineup(lineup));
    const totalLineup = showTotalLineups ? [
      // Have to do this last in order to get the luck-mutated lineups
      // TODO: luck adjusted lineups seemed completely wrong so I pulled it :(, see above
      // (https://github.com/Alex-At-Home/cbb-on-off-analyzer/issues/100)
      enrichLineup(_.assign(LineupUtils.calculateAggregatedLineupStats(filteredLineups), {
        key: LineupTableUtils.totalLineupId,
        doc_count: filteredLineups.length //(for doc_count >0 checks, calculateAggregatedLineupStats doesn't inject)
      }))
    ] : [];

    return totalLineup.concat(enrichedLineups);
  }

  /** Builds the list of where players play based on their lineup */
  static getPositionalInfo(
    lineups: Array<LineupStatSet>,
    positionFromPlayerId: Record<PlayerId, IndivPosInfo>,
    teamSeasonLookup: string
  ): PositionInfo[][] {
    return _.chain(lineups).transform((mutableAcc, lineup) => {
      const codesAndIds = LineupTableUtils.buildCodesAndIds(lineup);
      const sortedCodesAndIds = PositionUtils.orderLineup(codesAndIds, positionFromPlayerId, teamSeasonLookup);
      sortedCodesAndIds.forEach((codeId, i) => {
        mutableAcc[i]!.push({ id: codeId.id, code: codeId.code, numPoss: lineup?.off_poss?.value || 0 })
      });
    }, [
      [] as PositionInfo[], [] as PositionInfo[], [] as PositionInfo[], [] as PositionInfo[], [] as PositionInfo[],
    ]).map(keyPosses => {
      return _.chain(keyPosses).groupBy(keyPoss => keyPoss.id).mapValues((keyPosses, id) => {
        const code = keyPosses?.[0].code;
        return {
          id: id,
          code: code,
          numPoss: _.reduce(keyPosses, (acc, keyPoss) => acc + keyPoss.numPoss, 0)
        };
      }).values().orderBy([ "numPoss" ], [ "desc" ]).value();
    }).value();
  }

}
