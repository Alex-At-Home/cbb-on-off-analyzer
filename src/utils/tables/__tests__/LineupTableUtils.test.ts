import { LineupTableUtils } from '../LineupTableUtils';
import _ from "lodash";

import { SampleDataUtils } from "../../../sample-data/SampleDataUtils";
import { sampleLineupStatsResponse } from "../../../sample-data/sampleLineupStatsResponse";
import { samplePlayerStatsResponse } from "../../../sample-data/samplePlayerStatsResponse";
import { sampleTeamStatsResponse } from "../../../sample-data/sampleTeamStatsResponse";

describe("LineupTableUtils", () => {

  test("getPositionalInfo", () => {
    // Setup test data:
    const teamSeasonLookup = "Men_Maryland_2018/9";

    const lineups =
      sampleLineupStatsResponse.responses[0].aggregations.lineups.buckets || [];

    const globalRosterStats =
      samplePlayerStatsResponse.responses[0].aggregations?.tri_filter?.buckets?.baseline?.player?.buckets || [];

    const positionFromPlayerKey = LineupTableUtils.buildPositionPlayerMap(globalRosterStats, teamSeasonLookup);

    const results = LineupTableUtils.getPositionalInfo(
      lineups, positionFromPlayerKey, teamSeasonLookup
    );
    //expect(results).toBe({});
    expect(results).toMatchSnapshot();

  });
});
