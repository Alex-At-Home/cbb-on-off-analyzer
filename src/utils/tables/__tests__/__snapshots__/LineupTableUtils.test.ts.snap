// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineupTableUtils getPositionalInfo 1`] = `
Array [
  Array [
    Object {
      "key": "Cowan, Anthony",
      "numPoss": 809,
    },
  ],
  Array [
    Object {
      "key": "Wiggins, Aaron",
      "numPoss": 809,
    },
  ],
  Array [
    Object {
      "key": "Smith, Jalen",
      "numPoss": 809,
    },
  ],
  Array [
    Object {
      "key": "Ayala, Eric",
      "numPoss": 598,
    },
    Object {
      "key": "Scott, Donta",
      "numPoss": 211,
    },
  ],
  Array [
    Object {
      "key": "Morsell, Darryl",
      "numPoss": 612,
    },
    Object {
      "key": "Scott, Donta",
      "numPoss": 197,
    },
  ],
]
`;
